libraries/AP_HAL_SITL/DSP.cpp

```diff
@@ -0,0 +1,252 @@
+    uint16_t bin_range = (end_bin - start_bin) + 1;
+    max_f32(&fft->_freq_bins[start_bin], bin_range, &max_value, &fft->_max_energy_bin);
+    fft->_max_energy_bin += start_bin;
+    // scale the power to account for the input window
+    scale_f32(fft->_freq_bins, fft->_window_scale, fft->_freq_bins, fft->_bin_count);
+    fft->_update_step++;
+}
+
+// step 4: find the bin with the highest energy and interpolate the required frequency
+uint16_t DSP::step_calc_frequencies(FFTWindowStateSITL* fft, uint16_t start_bin, uint16_t end_bin)
+{
+    if (is_zero(fft->_freq_bins[fft->_max_energy_bin])) {
+        fft->_max_bin_freq = start_bin * fft->_bin_resolution;
+    } else {
+        // It turns out that Jain is pretty good and works with only magnitudes, but Candan is significantly better
+        // if you have access to the complex values and Quinn is a little better still. Quinn is computationally
+        // more expensive, but compared to the overall FFT cost seems worth it.
+        fft->_max_bin_freq = (fft->_max_energy_bin + calculate_quinns_second_estimator(fft, fft->_rfft_data, fft->_max_energy_bin)) * fft->_bin_resolution;
+    }
+    fft->_update_step++;
+
+    return fft->_max_energy_bin;
+}
+
+void DSP::mult_f32(const float* v1, const float* v2, float* vout, uint32_t len)
+{
+    for (uint32_t i = 0; i < len; i++) {
+        vout[i] = v1[i] * v2[i];
+    }
+}
+
+void DSP::max_f32(const float* vin, uint32_t len, float* maxValue, uint32_t* maxIndex)
+{
+    *maxValue = vin[0];
+    *maxIndex = 0;
+    for (uint32_t i = 1; i < len; i++) {
+        if (vin[i] > *maxValue) {
+            *maxValue = vin[i];
+            *maxIndex = i;
+        }
+    }
+}
+
+void DSP::scale_f32(const float* vin, float scale, float* vout, uint32_t len)
+{
+    for (uint32_t i = 0; i < len; i++) {
+        vout[i] = vin[i] * scale;
+    }
+}
+
+// simple integer log2
+static uint32_t fft_log2(uint32_t n)
+{
+    uint32_t k = n, i = 0;
+    while (k) {
+        k >>= 1;
+        i++;
+    }
+    return i - 1;
+}
+
+// bitreverse the input
+static uint32_t fft_bitreverse(uint32_t length, uint32_t n)
+{
+    uint32_t p = 0;
+    for (uint32_t j = 1; j <= fft_log2(length); j++) {
+        if (n & (1 << (fft_log2(length) - j))) {
+            p |= 1 << (j - 1);
+        }
+    }
+    return p;
+}
+
+// reorder all elements of the input according to their bitreversed index
+static void fft_order(complexf *f1, uint32_t length, complexf* temp)
+{
+    for (uint32_t i = 0; i < length; i++) {
+        temp[i] = f1[fft_bitreverse(length, i)];
+    }
+    for (uint32_t j = 0; j < length; j++) {
+        f1[j] = temp[j];
+    }
+}
+
+// calculate the in-place FFT of the input using the Cooley–Tukey algorithm
+void DSP::calculate_fft(complexf *f, uint32_t length)
+{
+    complexf *temp = new complexf[length];
+    fft_order(f, length, temp);
+
+    temp[1] = std::polar(1.0f, -2.0f * M_PI / length);
+    temp[0] = 1;
+    for (uint32_t i = 2; i < length / 2; i++) {
```

Comment:
add a temp variable for lenght/2 before the for loop and a calculation

Reply:
Again, SITL so don't care that much :)